src/contracts/examples/fastlane-online/OuterHelpers.sol

```diff
@@ -127,46 +157,121 @@ contract OuterHelpers is FastLaneOnlineInner {
-            grossGasRefund += S_aggCongestionBuyIn[userOpHash];
+            if (solversSuccessful) {
+                _grossGasRefund += _congestionBuyIns;
+            }
             delete S_aggCongestionBuyIn[userOpHash];
         }
 
-        uint256 _netRake = grossGasRefund * _CONGESTION_RAKE / _CONGESTION_BASE;
+        uint256 _netRake = _grossGasRefund * _CONGESTION_RAKE / _CONGESTION_BASE;
 
         // Increment cumulative rake
-        rake += _netRake;
+        if (solversSuccessful) {
+            rake += _netRake;
+        } else {
+            // NOTE: We do not refund the congestion buyins to the user because we do not want to create a
+            // scenario in which the user can profit from Solvers failing. We also shouldn't give these to the
+            // validator for the same reason.
+            // TODO: _congestionBuyIns to protocol guild or something because contract authors should be credibly
+            // neutral too
+            rake += (_netRake + _congestionBuyIns);
+        }
 
         // Return the netGasRefund
-        netGasRefund = grossGasRefund - _netRake;
+        netGasRefund = _grossGasRefund - _netRake;
+    }
+
+    function _sortSolverOps(SolverOperation[] memory unsortedSolverOps)
+        internal
+        view
+        returns (SolverOperation[] memory sortedSolverOps)
+    {
+        // This could be made more gas efficient
+
+        uint256 _length = unsortedSolverOps.length;
+        if (_length == 0) {
+            return unsortedSolverOps;
+        }
+
+        sortedSolverOps = new SolverOperation[](_length);
+
+        uint256 _topBidAmount;
+        uint256 _topBidIndex;
+        bool _matched;
+
+        for (uint256 i; i < _length; i++) {
+            _topBidAmount = 0;
+            _topBidIndex = 0;
+            _matched = false;
+
+            for (uint256 j; j < _length; j++) {
+                uint256 _bidAmount = unsortedSolverOps[j].bidAmount;
+
+                if (_bidAmount >= _topBidAmount && _bidAmount != 0) {
+                    _topBidAmount = _bidAmount;
+                    _topBidIndex = j;
+                    _matched = true;
+                }
+            }
+
+            if (_matched) {
+                // Get the highest solverOp and add it to sorted array
+                SolverOperation memory _solverOp = unsortedSolverOps[_topBidIndex];
+                sortedSolverOps[i] = _solverOp;
+
+                // Mark it as sorted in old array
+                unsortedSolverOps[_topBidIndex].bidAmount = 0;
+            }
+        }
+
+        return sortedSolverOps;
+    }
+
+    function _updateSolverReputation(
+        SolverOperation[] memory solverOps,
+        uint128 magnitude,
+        bool solversSuccessful
+    )
+        internal
+    {
+        uint256 _length = solverOps.length;
+        for (uint256 i; i < _length; i++) {
+            if (solversSuccessful) {
+                S_solverReputations[solverOps[i].from].successCost += magnitude;
```

Comment:
This adds `magnitude` to `successCost` of all solvers involved in the solved metacall. I guess this is because its difficult to identify the winning solver at this point. Maybe `address winningSolver` is something we should be returning from a successful metacall.

Reply:
Ah we actually don't need to return it - we can just trigger it with a postSearcherOp.

Thats a good idea.

Reply:
oh nice, yeah thats an interesting way to do it. Will take a stab